### PR TITLE
Total time updaters

### LINKED ISSUE #321: Deprecate different signatures for updaters
### The problem

This is part of `mobject:mobject.update()` method:

https://github.com/ManimCommunity/manim/blob/505470010995bc5f8b3c3bc33d6636699784f566/manim/mobject/mobject.py#L155-L163

L159 fetches the parameters of **each** updater, **each time the object is updated**, which may occur multiple times in a single animation, let alone a whole scene. Fetching the parameter list makes use of the `inspect` library, which is not the fastest thing in the world. 

I think this is super unnecessary and adding a lot of overhead. This checking of the parameter list is done so that each updater can be called appropriately: either `updater(self, dt)`, or just `updater(self)`. The gain in convenience for the end user seems very small in comparison to the overhead that this incurs in.

### A possible fix

My suggestion is that we should just make each updater always accept two arguments, and never check the parameter list. The chunk of code above could just become
```python
    def update(self, dt=0, recursive=True):
        if self.updating_suspended:
            return self
        for updater in self.updaters:
            updater(self, dt)
```
No parameter fetching necessary.

Incidentally, the current version forces each updater function to name its second parameter `dt`, if it has one. With my proposal, updater functions' second argument may be called whatever the user wants. If an updater function doesn't need to use the second argument, they can just ignore it.

### Example

Here's a scene.

```python
from manim import Scene, Integer, VGroup, Transform

def dummy_updater(obj, dt):
    return [i**2 for i in range(100)]

class TestScene(Scene):
    def construct(self):
        number = Integer(0)
        for _ in range(1000):
            number.add_updater(dummy_updater)

        self.add(number)
        for i in range(10):
            self.play(Transform(number, Integer(i)))
```

When executed with the current master, I get

```bash
$ rm -r media/
$ time manim -l scene.py 
...suppressed manim output...
real	1m7.297s
user	0m57.463s
sys	0m2.940s
```

When ran with my suggestion:
```bash
$ rm -r media/
$ time manim -l scene.py 
...suppressed manim output...
real	0m43.912s
user	0m40.581s
sys	0m2.781s
```

### PR DESCRIPTION
**Pull request by AntonBallmaier**
## Changelog / Overview

- `Mobject.add_updater` has a new parameter `pass_relative_times` determining if the updater should get passed relative or absolute time values.
- Updaters can now have arbitraty named second parameters instead of requiring `dt`.
- `call_updater=True` no longer causes an error with non-time-based updaters.
- Renamed `Mobject.update` to `Mobject._apply_updaters` and deprecated `update`.

## Motivation
From time to time you don't need the time difference since last updater call, but the total time... Now you without a problem.
Closes #321
Includes parts of the feature requested in #851

## Explanation for Changes
[Documentation Reference](https://manimce--1520.org.readthedocs.build)

## Checklist
- [x] I have read the [Contributing Guidelines](https://docs.manim.community/en/latest/contributing.html)
- [x] I have written a descriptive PR title (see top of PR template for examples)
- [x] I have written a changelog entry for the PR or deem it unnecessary
- [x] My new functions/classes either have a docstring or are private
- [x] My new functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [x] My new documentation builds, looks correctly formatted, and adds no additional build warnings

## Reviewer Checklist
- [x] The PR title is descriptive enough
- [x] The PR is labeled correctly
- [x] The changelog entry is completed if necessary
- [ ] Newly added functions/classes either have a docstring or are private
- [ ] Newly added functions/classes have [tests](https://github.com/ManimCommunity/manim/wiki/Testing) added and (optional) examples in the docs
- [ ] Newly added documentation builds, looks correctly formatted, and adds no additional build warnings

### CODE REVIEW

File: manim/mobject/mobject.py [hunk 1]
```diff
@@ -866,24 +844,38 @@ def get_family_updaters(self):
...
+            named ``t`` or ``time``.
         index
-            The index at which the new updater should be added in ``self.updaters``. In case ``index`` is ``None`` the updater will be added at the end.
+            The index at which the new updater should be added in the list of updaters.
+            The order of this list determines the execution order of the updaters.
+            In case ``index`` is ``None`` the updater will be added at the end, where it
+            would be last in the execution order.
         call_updater
-            Whether or not to call the updater initially. If ``True``, the updater will be called using ``dt=0``.
+            Wheather or not to call the updater initially after adding it (using ``0``
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:877]:
```suggestion
            Whether or not to call the updater initially after adding it (using ``0``
```

File: manim/mobject/mobject.py [hunk 2]
```diff
@@ -925,22 +917,47 @@ def construct(self):
         :class:`~.UpdateFromFunc`
         """
 
+        # Test if updater is time based and wheather to use time difference
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:920]:
```suggestion
        # Test if updater is time based and whether to use time difference
```

File: manim/mobject/opengl_mobject.py [hunk 3]
```diff
@@ -612,26 +619,63 @@ def get_updaters(self):
...
-            updater_list = self.time_based_updaters
+    def add_updater(
+        self,
+        updater: Updater,
+        *,
+        pass_relative_times: Optional[bool] = None,
+        index: Optional[int] = None,
+        call_updater: bool = False,
+    ) -> "OpenGLMobject":
+        # Test if updater is time based and wheather to use time difference
```
Comment (jsonvillanueva) [inline manim/mobject/opengl_mobject.py:630]:
```suggestion
        # Test if updater is time based and whether to use time difference
```

**Review by jsonvillanueva — COMMENTED** [review]
pre-commit is complaining about these -- applying these changes

File: manim/mobject/mobject.py [hunk 4]
```diff
@@ -925,22 +917,47 @@ def construct(self):
         :class:`~.UpdateFromFunc`
         """
 
+        # Test if updater is time based and whether to use time difference
+        parameters = list(get_parameters(updater).keys())
+        time_based = len(parameters) > 1
+        if time_based:
+            if parameters[1] not in ["t", "time"] and pass_relative_times is None:
+                pass_relative_times = True
+            pass_relative_times = bool(pass_relative_times)
```
Comment (jsonvillanueva) [inline manim/mobject/mobject.py:926]: again, I don't particularly like this hard coded check for the second argument, but with a default of `True`, you can simplify the logic to only check for if it should be set to `False`
```suggestion
            if parameters[1] in ["t", "time"]:
                pass_relative_times = False
```

**Review by jsonvillanueva — APPROVED** [review]
In general, this PR functions as intended from my tests/usage of the new feature. I'll approve now as it works and this *could* be added to the 0.7.0 release, but I've offered some suggestions to consider.

**Comment by Darylgolden** [comment]
Hi, are you still planning on working on this?

**Comment by MrDiver** [comment]
We will come back to this with a fresh view if the issue get's relevant again. This PR is already linked in the original issue

- https://github.com/ManimCommunity/manim/issues/321

So i will close this for now. Feel free to open this PR again or a new PR which fixes the issue if you want to work on it again!

**Pull request closed without merging**